crates/conjure_core/src/ast/expressions.rs

```diff
@@ -567,6 +567,25 @@ impl Expression {
                 | Expression::Product(_, _)
         )
     }
+
+    /// True iff self and other are both atomic and identical.
+    ///
+    /// This method is useful to cheaply check equivalence. Assuming CSE is enabled, any unifiable
+    /// expressions will be rewritten to a common variable. This is much cheaper than checking the
+    /// entire subtrees of `self` and `other`.
+    pub fn identical_atom_to(&self, other: &Expression) -> bool {
```

Comment:
would this be cleaner like follows (untested)

```
    match (self.try_into(), other.try_into()) {
        (Ok(atom1), Ok(atom2)) => atom1 == atom2,
        _ => false,
    }
```

Reply:
I cannot inline the try_intos into matches/ if-lets as I can't give it the type annotations it needs that way...

Might be possible to clean it up using TryInto<&Atom>::try_into(&self), but I'm not sure if that is cleaner than what I have...

Reply:
You are right, I am still not sure (without trying) what Rust likes and what it doesn't.

You can keep it as is, or I would probably write the following if it was me. Maybe it's the Haskell influence in my style.

```
    let atom1: Result<&Atom, _> = self.try_into();
    let atom2: Result<&Atom, _> = other.try_into();

    match (atom1, atom2) {
        (Ok(atom1), Ok(atom2)) => atom1 == atom2,
        _ => false,
    }
```

Reply:
Took your suggestion, albeit using `if let` instead (clippy says to use it for single case matches)